src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SecurityChannelFactory.cs

```diff
@@ -0,0 +1,843 @@
+            return (TChannel)(object)new SecurityRequestSessionChannel(this, _securityProtocolFactory, ((IChannelFactory<IRequestSessionChannel>)InnerChannelFactory).CreateChannel(address, via), address, via);
+        }
+
+        protected override void OnOpen(TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            OnOpenCore(timeoutHelper.RemainingTime());
+            base.OnOpen(timeoutHelper.RemainingTime());
+            SetBufferManager();
+        }
+
+        protected internal override Task OnOpenAsync(TimeSpan timeout)
+        {
+            return OnOpenAsyncInternal(timeout);
+        }
+
+        private async Task OnOpenAsyncInternal(TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            await OnOpenCoreAsync(timeoutHelper.RemainingTime());
+            await base.OnOpenAsync(timeoutHelper.RemainingTime());
+            SetBufferManager();
+        }
+
+        void SetBufferManager()
+        {
+            ITransportFactorySettings transportSettings = GetProperty<ITransportFactorySettings>();
+
+            if (transportSettings == null)
+                return;
+
+            BufferManager bufferManager = transportSettings.BufferManager;
+
+            if (bufferManager == null)
+                return;
+
+            if (SessionMode && SessionClientSettings != null && SessionClientSettings.SessionProtocolFactory != null)
+            {
+                SessionClientSettings.SessionProtocolFactory.StreamBufferManager = bufferManager;
+            }
+            else
+            {
+                ThrowIfProtocolFactoryNotSet();
+                _securityProtocolFactory.StreamBufferManager = bufferManager;
+            }
+        }
+
+
+        protected override IAsyncResult OnBeginOpen(TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            throw ExceptionHelper.PlatformNotSupported(); // Issue #31 in progress
+            //return new OperationWithTimeoutAsyncResult(new OperationWithTimeoutCallback(OnOpen), timeout, callback, state);
+        }
+
+        protected override void OnEndOpen(IAsyncResult result)
+        {
+            throw ExceptionHelper.PlatformNotSupported(); // Issue #31 in progress
+            //OperationWithTimeoutAsyncResult.End(result);
+        }
+
+        void OnOpenCore(TimeSpan timeout)
+        {
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            if (SessionMode)
+            {
+                SessionClientSettings.Open(this, InnerChannelFactory, ChannelBuilder, timeoutHelper.RemainingTime());
+            }
+            else
+            {
+                ThrowIfProtocolFactoryNotSet();
+                _securityProtocolFactory.Open(true, timeoutHelper.RemainingTime());
+            }
+        }
+
+        private async Task OnOpenCoreAsync(TimeSpan timeout)
```

Comment:
>private [](start = 8, length = 7)

Question only: Do the new Open/CloseAsyncs need to be private? Or is this to prevent us from internally calling these methods by accident at the moment?